src/library/crate/cratefeature.cpp

```diff
@@ -0,0 +1,915 @@
+        // populate a list of files of the crate
+        QList<QString> playlist_items;
+        int rows = pCrateTableModel->rowCount();
+        for (int i = 0; i < rows; ++i) {
+            QModelIndex index = m_crateTableModel.index(i, 0);
+            playlist_items << m_crateTableModel.getTrackLocation(index);
+        }
+
+        if (file_location.endsWith(".pls", Qt::CaseInsensitive)) {
+            ParserPls::writePLSFile(file_location, playlist_items, useRelativePath);
+        } else if (file_location.endsWith(".m3u8", Qt::CaseInsensitive)) {
+            ParserM3u::writeM3U8File(file_location, playlist_items, useRelativePath);
+        } else {
+            //default export to M3U if file extension is missing
+            if(!file_location.endsWith(".m3u", Qt::CaseInsensitive))
+            {
+                qDebug() << "Crate export: No valid file extension specified. Appending .m3u "
+                         << "and exporting to M3U.";
+                file_location.append(".m3u");
+            }
+            ParserM3u::writeM3UFile(file_location, playlist_items, useRelativePath);
+        }
+    }
+}
+
+void CrateFeature::slotExportTrackFiles() {
+    // Create a new table model since the main one might have an active search.
+    QScopedPointer<CrateTableModel> pCrateTableModel(
+        new CrateTableModel(this, m_pTrackCollection));
+    pCrateTableModel->selectCrate(m_crateTableModel.selectedCrate());
+    pCrateTableModel->select();
+
+    int rows = pCrateTableModel->rowCount();
+    QList<TrackPointer> trackpointers;
+    for (int i = 0; i < rows; ++i) {
+        QModelIndex index = m_crateTableModel.index(i, 0);
+        trackpointers.push_back(m_crateTableModel.getTrack(index));
+    }
+
+    TrackExportWizard track_export(nullptr, m_pConfig, trackpointers);
+    track_export.exportTracks();
+}
+
+void CrateFeature::slotCrateTableChanged(CrateId crateId) {
+    m_lastRightClickedIndex = rebuildChildModel(crateId);
+}
+
+void CrateFeature::slotCrateContentChanged(CrateId crateId) {
+    QSet<CrateId> updatedCrateIds;
+    updatedCrateIds.insert(crateId);
+    updateChildModel(updatedCrateIds);
+}
+
+void CrateFeature::slotUpdateCrateLabels(const QSet<CrateId>& updatedCrateIds) {
+    updateChildModel(updatedCrateIds);
+}
+
+void CrateFeature::htmlLinkClicked(const QUrl& link) {
+    if (QString(link.path())=="create") {
+        slotCreateCrate();
+    } else {
+        qDebug() << "Unknown crate link clicked" << link;
+    }
+}
+
+void CrateFeature::slotTrackSelected(TrackPointer pTrack) {
+    m_pSelectedTrack = pTrack;
+
+    TreeItem* pRootItem = m_childModel.getRootItem();
+    VERIFY_OR_DEBUG_ASSERT(pRootItem != nullptr) {
+        return;
+    }
+
+    TrackId selectedTrackId =
+            pTrack ? pTrack->getId() : TrackId();
+
+    std::vector<CrateId> sortedTrackCrates;
+    CrateTrackSelectResult trackCratesIter(
+            m_pTrackCollection->crates().selectTrackCratesSorted(selectedTrackId));
```

Comment:
Alternatively you could use here the QSet returned by CrateStorage::collectCrateIdsOfTracks.  And use QSet::contains() below. Do you have a clue how this performs to the library sorting solution here? IMHO this would improve the readability.